tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc

```diff
@@ -1724,13 +1724,39 @@ class DynamicBroadcastInDimOpNotActuallyDynamic
                                 PatternRewriter& rewriter) const override {
     auto type = op.getType().dyn_cast<RankedTensorType>();
     auto operandType = op.operand().getType().dyn_cast<RankedTensorType>();
-    if (!type || !type.hasStaticShape() || !operandType ||
+    auto outputDimOp = op.output_dimensions().getDefiningOp();
+    auto outputDimType =
+        op.output_dimensions().getType().dyn_cast<RankedTensorType>();
+    if (!type || !operandType || !outputDimType ||
         !operandType.hasStaticShape()) {
-      return rewriter.notifyMatchFailure(op, "requires static shape");
+      return rewriter.notifyMatchFailure(op, "requires operand static shape");
     }
-    rewriter.replaceOpWithNewOp<BroadcastInDimOp>(
-        op, op.getType(), op.operand(), op.broadcast_dimensions());
-    return success();
+    if (type.hasStaticShape()) {
+      rewriter.replaceOpWithNewOp<BroadcastInDimOp>(op, type, op.operand(),
+                                                    op.broadcast_dimensions());
+      return success();
+    } else if (outputDimOp &&
+               outputDimOp->hasTrait<mlir::OpTrait::ConstantLike>()) {
+      SmallVector<int64_t, 4> outputShape;
+      DenseIntElementsAttr shapeAttr;
+      if (auto constOp = llvm::dyn_cast<mhlo::ConstOp>(outputDimOp)) {
+        shapeAttr = constOp.value().cast<DenseIntElementsAttr>();
```

Comment:
Can you write it this way: `  if (matchPattern(outputDimOp, m_Constant(&shapeAttr)) ` ; it avoids encoding the exact constant operation.

It should catch both cases.